src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -3821,6 +3820,276 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+            string name,
+            ITypeSymbol returnType,
+            ITypeSymbol leftType,
+            ITypeSymbol rightType)
+        {
+            var csharpReturnType = returnType.EnsureCSharpSymbolOrNull(nameof(returnType));
+            var csharpLeftType = leftType.EnsureCSharpSymbolOrNull(nameof(leftType));
+            var csharpRightType = rightType.EnsureCSharpSymbolOrNull(nameof(rightType));
+
+            // caller already checked all of these were not null.
+            Debug.Assert(csharpReturnType is not null);
+            Debug.Assert(csharpLeftType is not null);
+            Debug.Assert(csharpRightType is not null);
+
+            var syntaxKind = SyntaxFacts.GetOperatorKind(name);
+            if (syntaxKind == SyntaxKind.None)
+                throw new ArgumentException($"Illegal operator name '{name}'", nameof(name));
+
+            var binaryOperatorName = OperatorFacts.BinaryOperatorNameFromSyntaxKindIfAny(syntaxKind, SyntaxFacts.IsCheckedOperator(name));
+            if (binaryOperatorName != name)
+                throw new ArgumentException($"'{name}' was not a valid binary operator name", nameof(name));
+
+            // Lang specific checks to ensure this is an acceptable operator.
+            checkOperatorKinds();
+
+            return new SynthesizedIntrinsicOperatorSymbol(csharpLeftType, name, csharpRightType, csharpReturnType).GetPublicSymbol();
+
+            static bool isDynamicOrError(TypeSymbol type)
+                => type.TypeKind is TypeKind.Dynamic or TypeKind.Error;
+
+            void checkOperatorKinds()
+            {
+                // Dynamic built-in operators allow virtually all operations with all types.  So we do no further checking here.
+                if (isDynamicOrError(csharpReturnType) ||
+                    isDynamicOrError(csharpLeftType) ||
+                    isDynamicOrError(csharpRightType))
+                {
+                    return;
+                }
+
+                // Use fast-path check to see if this types are ok.
+                var binaryKind = OperatorFacts.SyntaxKindToBinaryOperatorKind(SyntaxFacts.GetBinaryExpression(syntaxKind));
+                var easyOutBinaryKind = OverloadResolution.BinopEasyOut.OpKind(binaryKind, csharpLeftType, csharpRightType);
+
+                if (easyOutBinaryKind != BinaryOperatorKind.Error)
+                {
+                    var signature = this.builtInOperators.GetSignature(easyOutBinaryKind);
+                    if (TypeSymbol.Equals(csharpReturnType, signature.ReturnType, TypeCompareKind.ConsiderEverything) &&
+                        TypeSymbol.Equals(csharpLeftType, signature.LeftType, TypeCompareKind.ConsiderEverything) &&
+                        TypeSymbol.Equals(csharpRightType, signature.RightType, TypeCompareKind.ConsiderEverything))
+                    {
+                        return;
+                    }
+                }
+
+                // bool operator ==(object, object) is legal.
```

Comment:
the following was me reimplementing the logic from BinaryOperatorOverloadResolution_NoEasyOut.  I couldn't find a way to do it in a cleaner fashion given taht that logic depends on BoundExpressions, not types.